trieste/utils/misc.py

```diff
@@ -462,3 +462,9 @@ def _flatten_module(  # type: ignore[no-untyped-def]
         for subvalue in subvalues:
             # Predicate is already tested for these values.
             yield subvalue
+
+
+def ensure_positive(x: TensorType) -> TensorType:
+    """Esure that all the elements in `x` are strictly positive (using a dtype-dependent
+    capping threshold."""
+    return tf.math.maximum(x, 1e-6 if x.dtype == tf.float32 else 1e-16)
```

Comment:
naive question, is 1e-6 the lowest we can have with single precision?

Reply:
Not at all. This was just based on scaling up the suggested value of 1e-16 for float64. Both numbers can go significantly smaller if we want: float32 can go down to aound 1e-38 and float64 to 2e-308. Do you have any intuition for how small we should make these?

Reply:
I think we may be fine with smallest number for each precision that makes it positive, though it may depend on the usage downstream - at the moment we are just taking sqrt and doing some multiplication, that will take it to equal 0 but in this use case it should be fine I think? eps contribution would be removed in these cases, but not sure if that's relevant

Reply:
we do want to avoid "jumps" so I would go with something close to the end of range - imagine having an adjacent point that is 1e-300 but then you swap 0 with 1e-32, that would create a jump, no?

Reply:
`ensure_positive` will turn both 0 and 1e-300 to 1e-32, so there won't be a jump (but there won't be any gradient either)